content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,530 @@
+
+#### `pull()`
+
+The dplyr code:
+
+```{r pull, eval = FALSE}
+pull(mtcars_tbl, hp)
+```
+
+...can be substituted with:
+
+```{r pull-alt, eval = FALSE}
+mtcars_tbl$hp
+```
+
+...or:
+
+```{r pull-alt-2, eval = FALSE}
+mtcars_tbl[["hp"]]
+```
+
+Note that `$` will raise a warning if the subsetted column doesn't exist, while `[[` will silently return `NULL`.
+
+With benchmarks:
+
+```{r bm-pull}
+mark(
+  dplyr      = pull(mtcars_tbl, hp),
+  `$.tbl_df` = mtcars_tbl$hp,
+  `[[.tbl_df` = mtcars_tbl[["hp"]]
+) %>%
+  select(expression, median)
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#935](https://github.com/tidymodels/parsnip/pull/935) and [tidymodels/tune#635](https://github.com/tidymodels/tune/pull/635).
+
+#### `bind_*()`
+
+`bind_rows()` and `bind_cols()` can be substituted for `vec_rbind()` and `vec_cbind()`, respectively. First, row-binding:
+
+```{r bm-bind-rows}
+mark(
+  dplyr = bind_rows(mtcars_tbl, mtcars_tbl),
+  vctrs = vec_rbind(mtcars_tbl, mtcars_tbl)
+) %>%
+  select(expression, median)
+```
+
+As for column-binding:
+
+```{r bm-bind-cols}
+tbl <- tibble(year = rep(1974L, nrow(mtcars_tbl)))
+
+mark(
+  dplyr = bind_cols(mtcars_tbl, tbl),
+  vctrs = vec_cbind(mtcars_tbl, tbl)
+) %>%
+  select(expression, median)
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/tune#636](https://github.com/tidymodels/tune/pull/636).
+
+#### `group_by()` + `summarize()`
+
+In general, the introduction of groups makes these substitutions much trickier. In those cases, it's likely best to weigh (via profiling) how significant the slowdown is and, if it's not too bad, opt not to make any changes. For code that relies on `group_by()` and sees heavy traffic, see `vctrs::list_unchop()`, `vctrs::vec_chop()`, and `vctrs::vec_rep_each()`.
+
+### Tibbles
+
+Tibbles are great, and I don't want to interface with any other data frame-y thing. Some notes:
+
+- `as_tibble()` on a tibble is not "free":
+
+```{r bm-as-tibble}
+mark(
+  on_tbl_df = as_tibble(mtcars_tbl),
+  on_data.frame = as_tibble(mtcars, rownames = "make_model")
+) %>% 
+  select(expression, median)
+```
+
+- Building a tibble from scratch using `tibble()` actually takes quite a while as well. `tibble()` handles vector recycling and name checking, builds columns sequentially, all that good stuff. If you need that, use `tibble()`, but if you're building a tibble from well-understood inputs, use `new_tibble(list())`. For a middle ground between `tibble()` and `new_tibble(list())` in terms of both performance and safety, use the `df_list()` function from the vctrs package in place of `list()`.
+
+```{r bm-tibble}
+mark(
+  tibble = tibble(a = 1:2, b = 3:4),
+  new_tibble_df_list = new_tibble(df_list(a = 1:2, b = 3:4), nrow = 2),
```

Comment:
I see that `df_list()` itself is super quick, but that `new_tibble()` `unclass()`ing its input results in most of that slowdown. Is this not a pattern we should be recommending?